sys/kern/pci.c

```diff
@@ -31,22 +31,41 @@ static const pci_vendor_id *pci_find_vendor(uint16_t vendor_id) {
 }
 
 static bool pci_device_present(device_t *pcid) {
-  return pci_read_config(pcid, PCIR_DEVICEID, 4) != -1U;
+  return pci_read_config_4(pcid, PCIR_DEVICEID) != -1U;
 }
 
 static int pci_device_nfunctions(device_t *pcid) {
+  /* If first function of a device is invalid, then
+   * no more functions are present. */
   if (!pci_device_present(pcid))
     return 0;
-  uint32_t hdrtype = pci_read_config(pcid, PCIR_HEADERTYPE, 1);
+  uint32_t hdrtype = pci_read_config_1(pcid, PCIR_HEADERTYPE);
   return (hdrtype & PCIH_HDR_MF) ? PCI_FUN_MAX_NUM : 1;
 }
 
-static uint32_t pci_bar_size(device_t *pcid, int bar, uint32_t addr) {
-  pci_write_config(pcid, PCIR_BAR(bar), 4, -1U);
-  uint32_t res = pci_read_config(pcid, PCIR_BAR(bar), 4);
+static uint32_t pci_bar_size(device_t *pcid, int bar, uint32_t *addr) {
+  /* Memory and I/O space accesses must be disabled via the
+   * command register before sizing a Base Address Register. */
+  uint16_t cmd = pci_read_config_2(pcid, PCIR_COMMAND);
+  pci_write_config_2(pcid, PCIR_COMMAND,
+                     cmd & ~(PCIM_CMD_MEMEN | PCIM_CMD_PORTEN));
+
+  /* Read the original value. */
+  uint32_t old = pci_read_config_4(pcid, PCIR_BAR(bar));
+  /* XXX: we don't handle 64-bit memory space bars. */
+
+  /* Write 0xFFFFFFFF and read it back. */
```

Comment:
I can read the code below, which states exactly what your code does. It does not explain why you do it.

Reply:
Done.